### PR TITLE
Fixes Implement google authentication with conflicts resolved #10 

### LINKED ISSUE #10: Implement google authentication
## Tasks:

- Implement google authentication (route should start with `/auth`)
- Add documentation

### PR DESCRIPTION
**Pull request by yamanrajab90**
Fixes #10

### CODE REVIEW

**Comment by yamanrajab90** [comment]
Fixes #10

File: swagger.json [hunk 1]
```diff
@@ -19,5 +19,81 @@
...
+        },
+        "name": {
+          "type": "string"
+        }
+      }
+    },
+    "Users": {
+      "type": "array",
+      "$ref": "#/definitions/User"
+    }
```
Comment (dilarafirtina) [inline swagger.json:97]: Great work!

**Review by dilarafirtina — APPROVED** [review]

File: swagger.json (hunk 1 above)
Comment (yamanrajab90) [reply swagger.json:454]: Thank you Dilara, the issue name you man on Github or something in the specefied lines in the Swagger file.

File: src/controllers/googleAuth.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+const jwt = require('jsonwebtoken');
+const { SECRET_KEY } = require('../middlewares/validation');
+require('../middlewares/passport');
+
+exports.authenticationCookie = async (req, res) => {
+  const { name, email, providerId, _id } = req.user;
+  const userInToken = {
```
Comment (dilarafirtina) [inline src/controllers/googleAuth.js:7]: naming can be "payload"

File: src/controllers/googleAuth.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    id: _id,
+    name,
+    email,
+    providerId: `google-${providerId}`,
+  };
```
Comment (dilarafirtina) [inline src/controllers/googleAuth.js:12]: _id:_id is enough for the payload data

File: src/controllers/googleAuth.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+  const token = jwt.sign(userInToken, SECRET_KEY, {
+    expiresIn: '336h',
```
Comment (dilarafirtina) [inline src/controllers/googleAuth.js:15]: expirey can be '14d'

File: src/controllers/googleAuth.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  });
+
+  res.cookie('token', token, {
+    httpOnly: true,
```
Comment (dilarafirtina) [inline src/controllers/googleAuth.js:19]: some fields are missing

File: src/controllers/googleAuth.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  });
+
+  res.redirect('/');
```
Comment (dilarafirtina) [inline src/controllers/googleAuth.js:22]: no need for redirect

File: src/controllers/googleAuth.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      name,
+      email,
+    };
+    res.status(200).json(clientUser);
+  };
+};
+exports.logout = async () => {
+  // eslint-disable-next-line no-unused-expressions, no-shadow
+  (req, res) => {
+    res.clearCookie('token').sendStatus(200).redirect('/');
```
Comment (dilarafirtina) [inline src/controllers/googleAuth.js:39]: redirect can be deleted

File: src/middlewares/passport.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+const GoogleStrategy = require('passport-google-oauth20').Strategy;
+
+const Volunteer = require('../models/volunteer');
+
+passport.use(
+  new GoogleStrategy(
+    {
+      clientID: process.env.GAPP_CLIENT_ID,
+      clientSecret: process.env.GAPP_CLIENT_SECRET,
+      callbackURL: 'http://localhost:3000/api/auth/google/callback',
```
Comment (dilarafirtina) [inline src/middlewares/passport.js:11]: url can be env value

File: src/middlewares/passport.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+          lastName: profile.json.family_name,
+          provider: profile.provider,
+        });
+
+        newVolunteer.save();
+        callBack(null, newVolunteer);
+      }
+    }
+  )
+);
```
Comment (dilarafirtina) [inline src/middlewares/passport.js:32]: great way to use oauth2!

File: src/middlewares/validation.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    req.user = loggedUser;
+    next();
+  } catch (error) {
+    res.clearCookie('token').sendStatus(401);
+  }
+};
+
+module.exports = {
+  checkAuth,
+  SECRET_KEY,
```
Comment (dilarafirtina) [inline src/middlewares/validation.js:20]: no need to export secret key

File: src/routes/auth.js [hunk 11]
```diff
@@ -1,7 +1,37 @@
...
+const googleAuthController = require('../controllers/googleAuth');
+const { checkAuth } = require('../middlewares/validation');
+
+require('../middlewares/passport');
+// It handle redirecting the user to google OAuth consent page.
+
+router.get(
+  '/google',
+  passport.authenticate('google', {
+    session: false,
```
Comment (dilarafirtina) [inline src/routes/auth.js:17]: good catch about session false!

File: src/routes/auth.js [hunk 12]
```diff
@@ -1,7 +1,37 @@
...
+);
+// It  sets an authentication cookie that holds the user JWT.
+router.get(
+  '/google/callback',
+  passport.authenticate('google', { session: false }),
+  googleAuthController.authenticationCookie
+);
+
+// Authenticated only endpoint that responds with a user object as specified in the user schema.
+router.get('/me', checkAuth, googleAuthController.loggedInUser);
```
Comment (dilarafirtina) [inline src/routes/auth.js:29]: me endpoint can be deleted

**Review by dilarafirtina — CHANGES_REQUESTED** [review]

**Pull request merged**